data/src/main/java/com/android254/data/repos/SessionsManager.kt

```diff
@@ -40,87 +38,143 @@ class SessionsManager @Inject constructor(
+        return combine(sessionsFlow, bookmarksFlow) { sessions, bookmarks ->
+            sessions.map { it.toDomainModel() }
+                .map { session ->
+                    session.copy(isBookmarked = bookmarks.map { it.sessionId }.contains(session.id))
                 }
-            )
+                .filter { session -> session.isBookmarked }
+        }.flowOn(ioDispatcher)
+    }
+
+    override fun fetchFilteredSessions(query: String): Flow<List<Session>> {
+        val filteredSessions = dao.fetchSessionsWithFilters(SimpleSQLiteQuery(query))
+        val bookmarksFlow = bookmarkDao.getBookmarkIds()
+        return combine(filteredSessions, bookmarksFlow) { sessions, bookmarks ->
+            sessions.map { session ->
+                session.copy(
+                    isBookmarked = bookmarks.map { it.sessionId }.contains(session.id.toString())
+                )
+            }.map { it.toDomainModel() }
+        }.flowOn(ioDispatcher)
+    }
+
+    override fun fetchSessionById(sessionId: String): Flow<Session?> {
+        val bookmarksFlow = bookmarkDao.getBookmarkIds()
+        val sessionFlow = dao.getSessionById(sessionId).map {
+            it?.toDomainModel()
+        }
+        return combine(sessionFlow, bookmarksFlow) { session, bookmarks ->
+            session?.copy(isBookmarked = bookmarks.map { it.sessionId }.contains(session?.id.toString()))
         }
     }
 
-    override suspend fun fetchSessionById(id: String): ResourceResult<Session> =
+    override suspend fun bookmarkSession(id: String) {
         withContext(ioDispatcher) {
-            val session = dao.getSessionById(id)
-                ?: return@withContext ResourceResult.Error(message = "requested event no longer available")
-            return@withContext ResourceResult.Success(data = session.toDomainModel())
+            bookmarkDao.insert(BookmarkEntity(id))
         }
+    }
 
-    override suspend fun toggleBookmarkStatus(
-        id: String,
-        isCurrentlyStarred: Boolean
-    ): ResourceResult<Boolean> = withContext(ioDispatcher) {
-        try {
-            dao.updateBookmarkedStatus(id, !isCurrentlyStarred)
-            if (isCurrentlyStarred) {
-                bookmarkDao.delete(BookmarkEntity(id))
-            } else {
-                bookmarkDao.insert(BookmarkEntity(id))
-            }
-        } catch (e: Exception) {
-            when (e) {
-                is NetworkError -> {
-                    return@withContext ResourceResult.Error("Network error")
-                }
-                else -> {
-                    return@withContext ResourceResult.Error("Error fetching sessions")
-                }
-            }
+    override suspend fun unBookmarkSession(id: String) {
+        withContext(ioDispatcher) {
+            bookmarkDao.delete(BookmarkEntity(id))
         }
-        return@withContext ResourceResult.Success(data = dao.getBookmarkStatus(id))
     }
+//    @RequiresApi(Build.VERSION_CODES.O)
```

Comment:
@chege4179 one last one, ca we not remove existing tests, want to merge this now